src/burnchains/burnchain.rs

```diff
@@ -1186,6 +1185,23 @@ impl Burnchain {
 
         // fetch all headers, no matter what
         let mut end_block = indexer.sync_headers(sync_height, None)?;
+        if did_reorg && sync_height > 0 {
+            // a reorg happened, and the last header fetched
+            // is on a smaller fork than the one we just
+            // invalidated. Wait for more blocks.
+            while end_block < db_height {
```

Comment:
my only concern is the use of a while loop here (in case the new blocks take a while to reach the `db_height`. It's my impression that elsewhere in the code base we "fail" and try again - for example, by returning `burnchain_error::TrySyncAgain`.  

I'm still approving, but wondering what your thoughts are on this.

Reply:
There are a few places in the codebase where we block on operations. 
This patch is clearly an edge case that we saw on Bitcoin Testnet (way more instable), and will probably never see on Mainnet. The TrySyncAgain was my initial approach, however on the next retry, we would be loosing the information "we just reorgd and waiting for more blocks from the new canonical fork", and would be trying to fetch blocks using the burnchain tip of the previous fork as a point of reference (which is the real issue).
This approach is limiting the footprint, I think the clean solution was introducing a DB schema modification, where we would update the BurnchainDB when we reorg, and marking the blocks from former canonical fork as such.
This approach is way more invasive and subject to introduce regression, which is something I'd like to avoid for the next few weeks :).